### ISSUE TITLE
DEBUG trap not honouring exit status 2

### ISSUE THREAD

**Issue by DesantBucie** [ksh93/ksh#187]
Hey again, I'm sorry that this isn't again related to code, but I cannot find any ksh forum or group. I'm running 93u+m.

I'm trying to make ksh function that **would run after every user command** and check input, to be more precise autocd.
The problem is i don't know how to do it, as you cannot probably alias regex, or at least i have failed.

I know there are built-in solutions in bash and zsh, but is it possible to be done on shell level?

**Comment by posguy99** [comment]
You can do $PROMPT_COMMAND in ksh with a discipline function, if that’s what you meant.

> function PS1.get 
{
	integer exitval=$?
	eval .sh.value=$($PROMPT_COMMAND)'$PS1'
	return $exitval
}

ref: https://www.mail-archive.com/ast-users@lists.research.att.com/msg01030.html

You could do all sorts of things with a getter on PS1.

--
Marc Wilson
posguy99@gmail.com

**Comment by McDutchie** [comment]
Interesting question. I wonder if it's possible.

`autocd` is a bash shell option (set using `shopt -s autocd`) that allows you to change into a directory by typing the directory name, without the need for a `cd` command.

When the prompt is displayed, the attempt to run the command is already past, and there is not really a way to get the command name to figure out if it's a directory. So I don't think a `PS1.get` discipline function is useful to implement something like `autocd`.

However, on ksh, the `DEBUG` trap may be useful... Says the manual:

> If *sig* is `DEBUG` then *action* will be executed before each command.  The variable `.sh.command` will contain the contents of the current command  line  when *action* is running.  If the exit status of the trap is 2 the command will not be executed.

So, a `DEBUG` trap could intercept the command, check if it's a directory, and if so, cd into it, without executing the command.

Let's make a little function...
```sh
_do_autocd()
{
        if      [[ -d $1 ]]
        then    printf 'cd -- %q\n' "$1"
                CDPATH= command cd -- "$1"
                return 2
        fi
}
```
Then let's set the trap to use it...
```sh
trap '_do_autocd "${.sh.command}"' DEBUG
```
Hmm. It works, it's changing directory alright, but it still tries to execute the directory as a command after `cd`ing, even though the function returns 2 in that case!

```
$ /usr/bin; echo "pwd is $PWD"
cd -- /usr/bin
-ksh: /usr/bin: cannot execute [Is a directory]
pwd is /usr/bin
$ ~; echo "pwd is $PWD"
cd -- /Users/martijn
-ksh: /Users/martijn: cannot execute [Is a directory]
pwd is /Users/martijn
$ /usr/local/src/ksh93/ksh; echo "pwd is $PWD"
cd -- /usr/local/src/ksh93/ksh
-ksh: /usr/local/src/ksh93/ksh: cannot execute [Is a directory]
pwd is /usr/local/src/ksh93/ksh
```

I wonder if we have just found another bug in the ksh `DEBUG` trap, because the manual clearly documents it's not supposed to execute the command if the `DEBUG` trap's exit status is 2, and that exit status should be inherited from the function call…

**Comment by McDutchie** [comment]
Sure enough, this function/trap combination works fine (without error messages) on every ksh version from 93t+ 2008-12-10 through to 93u+ 2012-08-01. It also works fine on ksh2020 which was a derivative of the 93v- beta (both abandoned now).

So, this is a bug introduced by ksh 93u+m. Ouch.

And this is why questions can be useful :)

I will try to hunt down this bug in the coming days. This autocd implementation should work.

**Comment by DesantBucie** [comment]
I have already tried trap, but only on u+m, and i still did it wrong. And in 5 minutes you wrote cleaner autocd function than all of my prototypes :D. Thanks, I'll try it on u+, and will be waiting for fix to test.

**Comment by DesantBucie** [comment]
By the way, can I use this function in [korny](https://github.com/DesantBucie/korny)? I will leave author and link to your profile either in docs or in code.

**Comment by posguy99** [comment]
The subdirectory is what makes it throw the error.  Why?  Couldn’t say.

> [ 3:53 PM][ttys003 +1][~]

> [131] mbp13 $ typeset -f _do_autocd                   

> _do_autocd()

>  {

>  if [[ -d "$1" ]]

>   then printf 'cd -- %q\n' "$1"

>   CDPATH= command cd -- "$1"

>    return 2

>   fi

> }

> [ 3:53 PM][ttys003 +1][~]

> [132] mbp13 $ trap '_do_autocd "${.sh.command}"' DEBUG

> [ 3:53 PM][ttys003 +1][~]

> [133] mbp13 $ /usr/bin; echo "pwd is $PWD"            

> cd -- /usr/bin

> -ksh: /usr/bin: cannot execute [Is a directory]

> pwd is /usr/bin

> [ 3:54 PM][ttys003 +1][/usr/bin]

> [134] mbp13 $ cd

> [ 3:54 PM][ttys003 +1][~]

> [135] mbp13 $ test; echo "pwd is $PWD"    

> cd -- test

> pwd is /Users/mwilson/test

> [ 3:54 PM][ttys003 +1][~/test]

> [136] mbp13 $ 


--
Marc Wilson
posguy99@gmail.com

**Comment by McDutchie** [comment]
> By the way, can I use this function in [korny](https://github.com/DesantBucie/korny)?

Sure, go right ahead.

**Comment by posguy99** [comment]
The subdirectory is what makes it throw the error.  Why?  Couldn’t say.

```
[ 3:53 PM][ttys003 +1][~]
[131] mbp13 $ typeset -f _do_autocd                   
_do_autocd()
 {
 if [[ -d "$1" ]]
  then printf 'cd -- %q\n' "$1"
  CDPATH= command cd -- "$1"
   return 2
  fi
}
[ 3:53 PM][ttys003 +1][~]
[132] mbp13 $ trap '_do_autocd "${.sh.command}"' DEBUG
[ 3:53 PM][ttys003 +1][~]
[133] mbp13 $ /usr/bin; echo "pwd is $PWD"            
cd -- /usr/bin
-ksh: /usr/bin: cannot execute [Is a directory]
pwd is /usr/bin
[ 3:54 PM][ttys003 +1][/usr/bin]
[134] mbp13 $ cd
[ 3:54 PM][ttys003 +1][~]
[135] mbp13 $ test; echo "pwd is $PWD"    
cd -- test
pwd is /Users/mwilson/test
[ 3:54 PM][ttys003 +1][~/test]
[136] mbp13 $ 
```

**Comment by McDutchie** [comment]
So, is the bug only triggered if the command name contains a slash? That would make the hunt more straightforward, but I don't think so:
```
$ lib; echo "$PWD"
cd -- lib
-ksh: lib: not found
/usr/local/src/ksh93/ksh/lib
```
In your case, the directory name was `test`, and that is a command. When executed without any arguments, that command quietly returns false (status 1). That's why you didn't see an error message.

**Comment by DesantBucie** [comment]
In my case error occurs, without slash to
<img width="216" alt="Screenshot 2021-02-20 at 01 10 40" src="https://user-images.githubusercontent.com/46130107/108574802-358c7300-7310-11eb-82d7-2dd47cc6c28a.png">

**Comment by McDutchie** [comment]
Commit d00b4b39 is the one that broke it.

So, the question is how to fix this bug without reintroducing the bug fixed in that commit (see the regression tests in that commit for examples).

**Comment by McDutchie** [comment]
This is the code that runs the DEBUG trap and (on line 1194-1195) avoids executing the command if the DEBUG trap yields exit status 2: https://github.com/ksh93/ksh/blob/4e47f89b0695a900e064cb88ba52179d058d9ea4/src/cmd/ksh93/sh/xec.c#L1182-L1196
The [sh_debug() function](https://github.com/ksh93/ksh/blob/4e47f89b0695a900e064cb88ba52179d058d9ea4/src/cmd/ksh93/sh/xec.c#L641-L709) calls sh_trap() in xec.c on line 696. The sh_trap() function is where the bug was introduced: https://github.com/ksh93/ksh/blob/4e47f89b0695a900e064cb88ba52179d058d9ea4/src/cmd/ksh93/sh/fault.c#L442-L502
Commit d00b4b39f66e2912797195f2aac9913218826350 removed the check for the debug trap from line 490 because it is faulty there.

However, to see if the DEBUG trap returned status 2, the xec.c code above checks the sh_trap()'s return value, as passed down via sh_debug(), and not the actual exit status variable. So to fix the bug, we do not need to reintroduce the faulty check before restoring the exit status. All we need to do is make sure that sh_trap() returns the exit status from *before* restoring the variable. Or at least that is my current hypothesis. Now to try it out...

**Comment by McDutchie** [comment]
The following patch fixes autocd for me, and causes zero regression test failures:
```diff
diff --git a/src/cmd/ksh93/sh/fault.c b/src/cmd/ksh93/sh/fault.c
index e50758c..915b7f1 100644
--- a/src/cmd/ksh93/sh/fault.c
+++ b/src/cmd/ksh93/sh/fault.c
@@ -442,11 +442,12 @@ void	sh_chktrap(Shell_t* shp)
 /*
  * parse and execute the given trap string, stream or tree depending on mode
  * mode==0 for string, mode==1 for stream, mode==2 for parse tree
+ * The return value is the exit status of the trap action.
  */
 int sh_trap(const char *trap, int mode)
 {
 	Shell_t	*shp = sh_getinterp();
-	int	jmpval, savxit = shp->exitval;
+	int	jmpval, savxit = shp->exitval, savxit_return;
 	int	was_history = sh_isstate(SH_HISTORY);
 	int	was_verbose = sh_isstate(SH_VERBOSE);
 	int	staktop = staktell();
@@ -487,6 +488,7 @@ int sh_trap(const char *trap, int mode)
 	sh_popcontext(shp,&buff);
 	shp->intrap--;
 	sfsync(shp->outpool);
+	savxit_return = shp->exitval;
 	if(jmpval!=SH_JMPEXIT && jmpval!=SH_JMPFUN)
 		shp->exitval=savxit;
 	stakset(savptr,staktop);
@@ -498,7 +500,7 @@ int sh_trap(const char *trap, int mode)
 	exitset();
 	if(jmpval>SH_JMPTRAP && (((struct checkpt*)shp->jmpbuffer)->prev || ((struct checkpt*)shp->jmpbuffer)->mode==SH_JMPSCRIPT))
 		siglongjmp(*shp->jmplist,jmpval);
-	return(shp->exitval);
+	return(savxit_return);
 }
 
 /*
```

**Comment by McDutchie** [comment]
xec.c, lines 1187-1192 are interesting: apparently, a DEBUG trap can set a 255 exit status to force return from a function. As far as I can tell, that's not documented anywhere.

**Comment by McDutchie** [comment]
Hey @DesantBucie, autocd has a bug in it: it doesn't work with directory names containing spaces or other characters requiring shellquoting. That's because, as it turns out, `${.sh.command}` automatically shell-quotes every argument, so we need an `eval` to remove the quotes and process the arguments properly. That wasn't documented anywhere, so it's going into the man page now.

Updated version, that now also works with two-argument `cd` invocations (`cd` *old* *new* substitutions, see `cd --man`):
```sh
_do_autocd()
{
	if	[[ -d "$1" ]]
	then	print -r "cd --$(printf ' %q' "$@")"
		CDPATH= command cd -- "$@"
		return 2
	fi
}
trap 'eval "_do_autocd ${.sh.command}"' DEBUG
```

**Comment by posguy99** [comment]
It’s documented in the man page, I don’t see it in the Bolsky anywhere.

If the exit status of the trap is 2 the command will not

              be executed.  If the exit status of the trap is 255 and inside a function or a dot script, the

              function or dot script will return.


--
Marc Wilson
posguy99@gmail.com

**Comment by McDutchie** [comment]
True, it's not in the book. The book also says that the `DEBUG` trap is executed after the command, which is wrong; it's executed before. Either this was always wrong or the behaviour changed not long after the book was written; the oldest versions of ksh93 I have available all run the DEBUG trap before the command and exit status 2 is a trigger to skip executing the next command.

**Comment by McDutchie** [comment]
Ping @DesantBucie 

Another bug in autocd: it killed the `.` command (that sources a dot script) because, of course, `[[ -d . ]]` is always true; the dot also refers to the current directory.

~~Fixed version:~~ -- nope, see below
```sh
_do_autocd()
{
        if      [[ -d $1 && $1 != '.' ]]
        then    print -r "cd --$(printf ' %q' "$@")"
                CDPATH= command cd -- "$@"
                return 2
        fi
}
trap 'eval "_do_autocd ${.sh.command}"' DEBUG
```

**Comment by avih** [comment]
> Another bug in autocd: it killed the . command

IMHO, and similarly it will cd instead of execute any command which also happens to be a dir (mkdir echo; echo). You should first check whether the thing can execute as a command...

**Comment by McDutchie** [comment]
Yup, good point. That's what bash seems to do, as well.

Easy enough:
```sh
_do_autocd()
{
        if      ! command -v "$1" >/dev/null && [[ -d $1 ]]
        then    print -r "cd --$(printf ' %q' "$@")"
                CDPATH= command cd -- "$@"
                return 2
        fi
}
trap 'eval "_do_autocd ${.sh.command}"' DEBUG
```

**Comment by McDutchie** [comment]
On a related note, I think the shell-quoted `${.sh.command}` implementation is a bit daft, really. They should just have made that into an indexed array, so we could have set the trap to `_do_autocd "${.sh.command[@]}"` instead of having to invoke `eval`. Oh well, can't change it now... maybe in future we can have a `${.sh.argv[@]}` array equivalent or some such.

**Comment by avih** [comment]
While argv could be useful, I think autocd should only get applied if the entire input is one thing (not argument, because it replaces the command) which is also a valid directory name (and cannot be executed as a command). Anything else (like `./foo bar` where `./foo` is a dir) - and the behavior should fallback to normal command IMO.

Also, I don't know where it gets applied, but IMHO it should only be effective when interactive, otherwise scripts could behave unexpectedly.

**Comment by McDutchie** [comment]
> I think autocd should only get applied if the entire input is one thing […]

Possibly. That's not how `shopt -s autocd` on bash behaves. `cd` on ksh actually supports two arguments:  "In the second form, the first occurrence of the string *old* contained in the pathname of the present working directory is replaced by the string *new* and the resulting string is used as the directory to which to change." However, that would make no sense in an autocd context. So I'm coming to the conclusion that you're right here.

This version does that:
```sh
_do_autocd()
{
	if	(($#==1)) && ! command -v "$1" >/dev/null && [[ -d $1 ]]
	then	printf 'cd -- %q\n' "$1" >&2
		CDPATH= command cd -- "$1"
		return 2
	fi
}
trap 'eval "_do_autocd ${.sh.command}"' DEBUG
```

This also prints the `cd --` tracing message to standard error instead of standard output, which makes more sense and is also what bash does.

By the way: a little-known ksh extension to tilde expansion is `~-`. It expands to `$OLDPWD` and works for any command like `~` and `~`*username* do. Since tilde expressions are expanded before command lookup, with autocd you can type `~-` to return to the previous directory.

> (not argument, because it replaces the command)

The command name/path is actually considered to be argument zero (`$0`).

> Also, I don't know where it gets applied, but IMHO it should only be effective when interactive, otherwise scripts could behave unexpectedly.

It gets applied wherever the user applies it. A typical use would be to put it in `~/.kshrc` which would only load it on interactive shells. I agree it makes no sense for scripts.

**Comment by avih** [comment]
> That's not how shopt -s autocd on bash behaves.

Right, but we're making the perfect version here ;)

> The command name/path is actually considered to be argument zero ($0).

Thanks. Is it referred to as argument zero in some context at POSIX (not that I recall...)? or are you just referring to implementation details?

**Comment by McDutchie** [comment]
> Is it referred to as argument zero in some context at POSIX (not that I recall...)?

It is referred to as argument zero all the way down to the level of the C language (`argv[0]`).

**Comment by DesantBucie** [comment]
Many thanks for updating the solution!

**Comment by avih** [comment]
> It is referred to as argument zero all the way down to the level of the C language (`argv[0]`).

Of course in C, but for shell command language POSIX states this:

> The words that are not variable assignments or redirections shall be expanded. If any fields remain following their expansion, the first field shall be considered the command name and remaining fields are the arguments for the command.

Not that it matters much, but I still don't recall that it's referred to as _argument_ zero by POSIX at the context of shell (obviously unrelated to the special parameter `$0`).

**Comment by McDutchie** [comment]
> > It is referred to as argument zero all the way down to the level of the C language (`argv[0]`).
> 
> Of course in C, but for shell command language POSIX states this:
> 
> > The words that are not variable assignments or redirections shall be expanded. If any fields remain following their expansion, the first field shall be considered the command name and remaining fields are the arguments for the command.

And then in *Command Execution* it states: "[…] the shell executes the utility in a separate utility environment (see Shell Execution Environment) with actions equivalent to calling the execl() function as defined in the System Interfaces volume of POSIX.1-2017 with the path argument set to the pathname resulting from the search, **arg0 set to the command name,** and the remaining execl() arguments set to the command arguments (if any) and the null terminator."

> Not that it matters much, but I still don't recall that it's referred to as _argument_ zero by POSIX at the context of shell (obviously unrelated to the special parameter `$0`).

It is very much related.

**Comment by DesantBucie** [comment]
Unfortunately last solution doesn't work on 2012 u+, nevertheless works beautiful on u+m.

**Comment by McDutchie** [comment]
> Unfortunately last solution doesn't work on 2012 u+, nevertheless works beautiful on u+m.

Hmm. It works for me on 93u+ 2012-08-01, but only if I use a simple `$PS1` prompt. With my usual complicated one that uses a `PS1.get` discipline function (a heavily tweaked version of @posguy99's prompt), ksh crashed rather spectacularly the first time I tried autocd, outputting garbage instead of the prompt and eventually segfaulting.

Further attempts simply ignored the `DEBUG` trap. I guess it's high time I got a 93u+m beta release out, 93u+ is nearly unusable... but even now, some of the unfixed bugs in 93u+m are just embarrassing.
